working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+  >  
+  > For example, you might have routes like `@books_bp.route("/<book_id>", methods=["GET", "PUT", "DELETE"])` where you would use `validate_book(book_id)` to ensure the provided book ID is valid before proceeding with the GET, PUT, or DELETE operation.
+  >  
+  > Please search your codebase for `validate_book` to find all its usages. In Visual Studio Code, you can do this by pressing `Ctrl+Shift+F` and typing `validate_book` in the search bar.
+</details>
+
+If we search our project, we have 4 dependencies: 
+- the `validate_book` function definition
+- usage in `read_one_book`
+- usage in `update_book`
+- usage in `delete_book`
+
+Peeking at our test suite in `test_routes.py`, we do have tests for `read_one_book`, but we need to write tests for the remaining untested dependencies: `validate_book`, `update_book`, and `delete_book`. 
+
+### Updating Our Test Suite
+
+#### Testing `validate_book`
+
+We'll start with testing the `validate_book` function. Before we begin, we're going to open several files to ensure Copilot has plenty of context around our code and test set up. Take a moment to ensure the following files are open in VS Code:
+- `book.py`
+- `routes.py`
+- `conftest.py`
+- `test_models.py`
+- `test_routes.py`
+
+We're going to start from `routes.py` and use our cursor to highlight the entire `validate_book` function. Next, bring up the inline Copilot chat and invoke the `/tests` shortcut. When Copilot is done thinking, we may see a slightly different UI than in previous scenarios:
+
+![A VS Code window with the routes.py file up and validate_book function selected showing a Refactor Preview tab at the bottom of the screen](assets/copilot-in-projects/routes-test-validate-book.png)  
+*Fig. Copilot's `Refactor Preview` tab showing at the bottom of VS Code*
+
+This time instead of showing changes to the current file in a temporary window, the bottom panel has opened to a `Refactor Preview` tab that has a single entry on it. If we click the entry, it will take us to a temporary file showing the changes Copilot is suggesting that we make to the `test_routes.py` file. 
+
+![A VS Code window with a temporary file showing that contains the original contents of test_routes.py along with suggested changes to test the `validate_book` function](assets/copilot-in-projects/test-routes-validate-book-refactor-preview.png)  
+*Fig. Temporary file in VS Code showing the test suggestions from Copilot*
+
+Here we can review the tests cases Copilot generated and see if there are changes or further tests we need. We have "Apply" and "Discard" controls at the bottom of the screen in the `Refactor Preview` tab to either accept or reject the changes.
+
+If we examine the code Copilot created, the scenarios identified are great – we have tests for the nominal case and a couple edge cases of invalid or non-existent book ids. However, there are a number of issues:
```

Comment:
- if we don't examine them then they aren't? schroedinger's code? 😁 
- code copilot created is a bit of a tongue twister
- semicolon for related independent clauses
```suggestion
Examining the code created by Copilot, the scenarios identified are great; we have tests for the nominal case and a couple edge cases of invalid or non-existent book ids. However, there are a number of issues:
```